tests/admin_app_tests/endpoint_tests/images_test.go

```diff
@@ -0,0 +1,222 @@
+	database_mock := mocks.DatabaseMock{
+		AddImageHandler: func(id string, file_name string, alt_text string) error {
+			return nil
+		},
+	}
+
+	r := admin_app.SetupRoutes(app_settings, database_mock)
+	w := httptest.NewRecorder()
+
+	pr, pw := io.Pipe()
+	writer := multipart.NewWriter(pw)
+
+	go func() {
+		defer writer.Close()
+
+		part, err := createTestForm(writer, "file", "test.png", "image/png")
+		if err != nil {
+			t.Error(err)
+		}
+
+		img := createImage()
+		if err != nil {
+			t.Error(err)
+		}
+
+		err = png.Encode(part, img)
+		if err != nil {
+			t.Error(err)
+		}
+	}()
+
+	req, _ := http.NewRequest("POST", "/images", pr)
+	req.Header.Add("Content-Type", writer.FormDataContentType())
+	r.ServeHTTP(w, req)
+
+	assert.Equal(t, 200, w.Code)
+
+	var response imageResponse
+	err := json.Unmarshal(w.Body.Bytes(), &response)
+	assert.Nil(t, err)
+
+	assert.NotNil(t, response.Id)
+}
+
+func TestPostImageNotAnImageFile(t *testing.T) {
+	database_mock := mocks.DatabaseMock{
+		AddImageHandler: func(id string, file_name string, alt_text string) error {
+			return nil
+		},
+	}
+
+	r := admin_app.SetupRoutes(app_settings, database_mock)
+	w := httptest.NewRecorder()
+
+	pr, pw := io.Pipe()
+	writer := multipart.NewWriter(pw)
+
+	go func() {
+		defer writer.Close()
+
+		part, err := createTestForm(writer, "file", "test.png", "image/png")
+		if err != nil {
+			t.Error(err)
+		}
+
+		text := bytes.NewBufferString("This is some dumy text to check the content test")
+		_, err = io.Copy(part, text)
+		if err != nil {
+			t.Error(err)
+		}
+	}()
+
+	req, _ := http.NewRequest("POST", "/images", pr)
+	req.Header.Add("Content-Type", writer.FormDataContentType())
+	r.ServeHTTP(w, req)
+
+	assert.Equal(t, 400, w.Code)
+}
+
+func TestPostImageWrongFileContentType(t *testing.T) {
+	database_mock := mocks.DatabaseMock{
+		AddImageHandler: func(id string, file_name string, alt_text string) error {
+			return nil
+		},
+	}
+
+	r := admin_app.SetupRoutes(app_settings, database_mock)
+	w := httptest.NewRecorder()
+
+	pr, pw := io.Pipe()
+	writer := multipart.NewWriter(pw)
+
+	go func() {
+		defer writer.Close()
+
+		part, err := createTestForm(writer, "file", "test.png", "application/json")
+		if err != nil {
+			t.Error(err)
+		}
+
+		img := createImage()
+
+		if err != nil {
+			t.Error(err)
+		}
+
+		err = png.Encode(part, img)
+		if err != nil {
+			t.Error(err)
+		}
+	}()
+
+	req, _ := http.NewRequest("POST", "/images", pr)
+	req.Header.Add("Content-Type", writer.FormDataContentType())
+	r.ServeHTTP(w, req)
+
+	assert.Equal(t, 400, w.Code)
+}
+
+func TestPostImageFailedToCreateDatabaseEntry(t *testing.T) {
+	database_mock := mocks.DatabaseMock{
+		AddImageHandler: func(id string, file_name string, alt_text string) error {
+			return errors.New("test error")
+		},
+	}
+
+	r := admin_app.SetupRoutes(app_settings, database_mock)
+	w := httptest.NewRecorder()
+
+	pr, pw := io.Pipe()
+	writer := multipart.NewWriter(pw)
+
+	go func() {
```

Comment:
Does this need to be running as a `goroutine` ? Sounds like we can do it all synchronously here right?

Reply:
Yeah, it kinda has to be a goroutine. The test don't work otherwise. :/

Reply:
I just tried to play with the `createPart(...)` code. I still don't understand why it needs to be on a separate thread (perhaps something to do with the writer requiring something from the reader pipe?) - But I couldn't get an easier solution to this without the goroutine.

Let's go with this.